### DISCUSSION TITLE
pyright can't locate some class members, even tho I know it's correct it keeps flagging as error.

### DISCUSSION THREAD

**Discussion by GabrielSousa02** [microsoft/pyright#6622] [Q&A] [answered]
I am using LazyVim with a python setup that uses `pyright`.

I was having some Django components errors, then installed `django-types` and that got it fixed, however now I'm facing some other erros, like the ones showed below:

![Screenshot 2023-12-02 at 12 35 02 AM](https://github.com/microsoft/pyright/assets/10780289/f9b7430c-940e-468b-a5a8-8bd957985153)

![Screenshot 2023-12-02 at 12 36 45 AM](https://github.com/microsoft/pyright/assets/10780289/dfbb021a-0b1e-4bf6-af76-844adf0a83f5)

I know these are false alerts mainly because tests are all passing, but also because some of the errors are on models I have created for the code, not only classes like `HttpResponse` or Django's user class.

Any light on where to go? I really wanted to remove those super annoying errors to be able to code on my Nvim setup at peace.

thank you.

**Comment by erictraut** [comment]
I can't really debug issues from partial screen shots. Please create a minimal, self-contained code sample and repro steps if you'd like help diagnosing the issue.

**Reply by GabrielSousa02** [reply]
Hey there Eric, sorry, I'll post the code. 

The reason I did not was because I thought this part was more like a forum than an Issue, so before reporting as a bug I considered asking for help to the community.

I'm getting the code to post it here.

I simply have a Django with APIs using DRF, and I created a custom user model and have some tests for the creation and other actions with that user. All tests are passing, but the error message is there.

*This is the model*
```python
"""
Database models.
"""

from django.db import models
from django.contrib.auth.models import AbstractBaseUser
from django.contrib.auth.models import BaseUserManager
from django.contrib.auth.models import PermissionsMixin


class UserManager(BaseUserManager):
    """Manager for Users"""

    def create_user(self, email, password=None, **extra_fields):
        """Create, save and return a new user."""
        if not email:
            raise ValueError("User must have an email address.")
        user = self.model(email=self.normalize_email(email), **extra_fields)
        user.set_password(password)
        user.save(using=self._db)

        return user

    def create_superuser(self, email, password):
        """Create and return a new superuser."""
        user = self.create_user(email=email, password=password)
        user.is_staff = True
        user.is_superuser = True
        user.save(using=self._db)
        return user


class User(AbstractBaseUser, PermissionsMixin):
    """User in the system."""

    email = models.EmailField(max_length=250, unique=True)
    name = models.TextField(max_length=250)
    last_name = models.TextField(max_length=250)
    is_active = models.BooleanField(default=True)
    is_staff = models.BooleanField(default=False)
    is_premium = models.BooleanField(default=False)
    password = models.TextField()

    objects = UserManager()

    USERNAME_FIELD = "email"
```

*This is my TEST for the model*
```python
"""
Test for Core models.
"""

from django.test import TestCase
from django.contrib.auth import get_user_model


class ModelTests(TestCase):
    """Test models."""

    def test_create_user_with_email_successful(self):
        """Test creating a user with an email is successful."""
        email = "test@example.com"
        password = "testpass123"
        user = get_user_model().objects.create_user(  # This is the line where the error appears
            email=email,
            password=password,
        )

        self.assertEqual(user.email, email)
        self.assertTrue(user.check_password(password))
```

This type of error also occurs on some other models/objects (e.g. the data returned on a request object, or the auto generated id of a custom model I created are flagged as unknown - even tho the tests are passing), however this reply would be huge if I post it all here. Please let me know if you would like to see it.

**Reply by erictraut** [reply, marked as answer]
Thanks. This isn't really a _minimal_, self-contained code sample ;), but it gives me something to go on.

If I understand your question correctly, you're wondering why pyright emits an error for the expression `get_user_model().objects.create_user`. 

* The expression `get_user_model()` evaluates to `type[AbstractBaseUser]`. This comes directly from the `django-stubs`, which has the following declaration: `def get_user_model() -> type[AbstractBaseUser]: ...`
* The expression `get_user_model().objects` evaluates to `BaseManager[AbstractUser]`. This comes from the definition of the `Model` class in the `django-stubs/db/models/base.pyi` file. 
```python
class Model(metaclass=ModelBase):
    ...
    objects: ClassVar[BaseManager[Self]]
```
* There is no `create_user` attribute or method on the `BaseManager[AbstractUser]` type. This is what pyright is telling you with this error.

You can do this sort of diagnosis yourself by using the `Go To Declaration` feature in your editor. This will take you to the place in the stubs that declare the type.

I don't know enough about django to understand how your custom `UserManager` and `User` classes are registered with the framework. I'm guessing it's some sort of dynamic registration? If so, a static type checker wouldn't know about these registrations.  You may need to resort to `cast` calls in that case.

```python
        user_manager = cast(UserManager, get_user_model().objects)
        user = user_manager.create_user(
            email=email,
            password=password,
        )
```

Perhaps some other pyright user knows enough about django to help you here. Or perhaps you'll need to post to a django forum to get a better answer to your question.

**Reply by GabrielSousa02** [reply]
I see. That explains it.

I kinda figure that it had something to do with that. But I'm fairly new to using pyright the way I'm using now, so I just wanted to make sure that I was not doing anything wrong and that it was a sort of false positive.

Thank you for the quick reply, and I'm sorry for the loads of information. I'm not used to using forums/dicussions, hence my lack of practice on providing information. Sorry, and again, thank you for your time.
